dev/tools/pre-commit

```
@@ -7,10 +7,25 @@ set -e
 
 dev/tools/check-overlays.sh
 
+# Can we check and fix formatting?
+ocamlformat=$(command -v ocamlformat || echo true)
+if [ "$ocamlformat" = true ]
+then
+    1>&2 echo "Warning: ocamlformat is not in path. Cannot check formatting."
+fi
+
+# Is the syntax correct?
+if ! git diff --cached --name-only -z | grep -E '.*\.mli?$' -z | xargs -0 $ocamlformat
+then
+    ocamlformat=true
+    1>&2 echo "Warning: some files are not syntactically valid. Formatting will not be checked."
+fi
```

This seems dangerous, I suspect that if we do
- commit a.ml and b.ml with b.ml having syntax error
- amend with fix to b.ml (not touching a.ml)

then the formatting for a.ml will not be fixed.

Why do you think ignoring errors from ocamlformat is worse?

---

Indeed, but that's the problem of the people whose commits do not compile. This is something that will be rejected by the linter anyway. Am I correct?

>Why do you think ignoring errors from ocamlformat is worse?

To get clear error messages. With your solution, you could get "Fixing..." even if nothing is fixed, whitespace nor formatting.

---

>To get clear error messages. With your solution, you could get "Fixing..." even if nothing is fixed, whitespace nor formatting.

Change it to "attempting to fix", problem solved?

---

I don't understand what is the issue with my solution and the fact that it is anyway a very edge case and not a big deal that some formatting is not fixed in this case.

---

>I don't understand what is the issue with my solution

We know how to do better.

Another scenario:
- `git add` a file with valid syntax
- break the file in the worktree
- `git commit`

The commit will have bad formatting.

---

OK, thanks, I understand the problem better with this scenario.

---

Actually @SkySkimmer the problem you describe in your scenario exists even without this preliminary check. Suppose I `git add` a file with broken syntax, then fix it in the worktree (including formatting), then `git commit` without option `-a`. The pre-commit hook will skip the fixing step and thus my commit will have bad formatting.